### ISSUE TITLE
OutOfBounds exception with NameResolvers and TransportSet

### ISSUE THREAD

**Issue by jhspaybar** [grpc/grpc-java#1657]
Long story short, it appears there's an explicit array.get(0) that happens on the first outbound call, but if the NameResolver hasn't found any servers yet, this causes an exception.  Stack trace can be found below(SpectatorInterceptor is a proprietary metrics component, and the EurekaNameResolver simply uses Eureka to resolve a set of IP + port that can then be used to load balance).  In the case below, the list of servers the EurekaNameResolver passes to onUpdate is empty.

```
SEVERE: Exception while executing runnable io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$2@7d8bd17e
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at java.util.Collections$UnmodifiableList.get(Collections.java:1309)
	at io.grpc.internal.TransportSet.scheduleConnection(TransportSet.java:197)
	at io.grpc.internal.TransportSet.obtainActiveTransport(TransportSet.java:169)
	at io.grpc.internal.ManagedChannelImpl$3.getTransport(ManagedChannelImpl.java:381)
	at io.grpc.SimpleLoadBalancerFactory$SimpleLoadBalancer$1.get(SimpleLoadBalancerFactory.java:121)
	at io.grpc.SimpleLoadBalancerFactory$SimpleLoadBalancer$1.get(SimpleLoadBalancerFactory.java:119)
	at io.grpc.internal.BlankFutureProvider$FulfillmentBatch.link(BlankFutureProvider.java:128)
	at io.grpc.SimpleLoadBalancerFactory$SimpleLoadBalancer.handleResolvedAddresses(SimpleLoadBalancerFactory.java:119)
	at io.grpc.internal.ManagedChannelImpl$2.onUpdate(ManagedChannelImpl.java:165)
	at com.netflix.grpc.nameresolver.eureka.EurekaNameResolver$1.onNext(EurekaNameResolver.java:64)
	at com.netflix.grpc.nameresolver.eureka.EurekaNameResolver$1.onNext(EurekaNameResolver.java:61)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:305)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:51)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:51)
	at com.netflix.grpc.interceptor.spectator.SpectatorMetricsClientInterceptor$1$1.onMessage(SpectatorMetricsClientInterceptor.java:44)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$2.runInContext(ClientCallImpl.java:423)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:54)
	at io.grpc.internal.SerializingExecutor$TaskRunner.run(SerializingExecutor.java:154)
	at java.util.concurrent.ThreadPoolExecutorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jhspaybar** [comment]
Worse yet, this appears to cause some real nastiness on the client interaction.  It appears the first call fails, and then the second attempted call on the same client blocks forever.

`
2016-04-11 10:50:55,382 ERROR com.netflix.alphagrpc3.grpc.ClientTest$2:123 [Thread-12] [run] Before call!
2016-04-11 10:50:55,393 ERROR com.netflix.alphagrpc3.grpc.ClientTest$2:138 [Thread-12] [run] Call failed
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at java.util.Collections$UnmodifiableList.get(Collections.java:1309)
	at io.grpc.internal.TransportSet.scheduleConnection(TransportSet.java:197)
	at io.grpc.internal.TransportSet.obtainActiveTransport(TransportSet.java:169)
	at io.grpc.internal.ManagedChannelImpl$3.getTransport(ManagedChannelImpl.java:381)
	at io.grpc.SimpleLoadBalancerFactory$SimpleLoadBalancer.pickTransport(SimpleLoadBalancerFactory.java:97)
	at io.grpc.internal.ManagedChannelImpl$1.get(ManagedChannelImpl.java:135)
	at io.grpc.internal.ClientCallImpl.start(ClientCallImpl.java:206)
	at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:45)
	at com.netflix.grpc.interceptor.request_context_bridge.RequestContextBridgeClientInterceptor$1.start(RequestContextBridgeClientInterceptor.java:122)
	at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:45)
	at com.netflix.grpc.interceptor.logging_context.LoggingContextClientInterceptor$1.start(LoggingContextClientInterceptor.java:34)
	at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:45)
	at com.netflix.grpc.interceptor.spectator.SpectatorMetricsClientInterceptor$1.start(SpectatorMetricsClientInterceptor.java:53)
	at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:45)
	at com.netflix.grpc.interceptor.salp.SalpClientInterceptor$1.start(SalpClientInterceptor.java:45)
	at io.grpc.stub.ClientCalls.startCall(ClientCalls.java:245)
	at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:225)
	at io.grpc.stub.ClientCalls.futureUnaryCall(ClientCalls.java:186)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:132)
	at com.netflix.alphagrpc3.protogen.AlphaGrpc3ServiceGrpc$AlphaGrpc3ServiceBlockingStub.sayHello(AlphaGrpc3ServiceGrpc.java:109)
	at com.netflix.alphagrpc3.grpc.ClientTun(ClientTest.java:130)
	at java.lang.Thread.run(Thread.java:745)
2016-04-11 10:51:00,393 ERROR com.netflix.alphagrpc3.grpc.ClientTest$2:123 [Thread-12] [run] Before call!
`

`
final Runnable r = new Runnable() {
      @Override
      public void run() {
        while(true) {
          try {
            Thread.sleep(50_000);
            LOG.error("Before call!");
            HelloReply reply = client.sayHello(HelloRequest.newBuilder().setName("BOB").build());
            LOG.error("Reply is: {}", reply.getMessage());
            try {
              Thread.sleep(10_000);
            } catch (InterruptedException e) {
              LOG.error("Interrupted", e);
            }
          } catch (Exception e) {
            LOG.error("Call failed", e);
          }
        }
      }
    };
    Thread t = new Thread(r);
    t.setDaemon(true);
    t.start();
`

**Comment by ejona86** [comment]
The second issue may be related to #1407 (I don't know; I've not looked into it yet).

**Comment by jhspaybar** [comment]
Based on the description it almost certainly is related, the exception bubbles up through the onUpdate method.

**Comment by zhangkun83** [comment]
@jhspaybar While we should probably fix #1407 to prevent further calls from being stuck, I wanted to know why your name resolver would call `onUpdate()` with an empty list.

**Comment by jhspaybar** [comment]
Mostly because it's being lazy :)  We have a Eureka API that just streams back a complete snapshot of IP+port pairs every ~90 seconds, and we essentially just responseList.map(this::toResolvedServerInfos) and pass it straight in to onUpdate.  For now, we're just further doing a size check and calling listener.onError(Status.UNAVAILABLE.withDescription("Eureka URI found no servers.")) when it's empty and only calling onUpdate when servers are there.  This seems to function well as a workaround.
